KSNM Infernal Swarm

dallano opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Implemented KSNM Infernal Swarm

Changes:
 - Gave Beelzebub a potent pollen ability (~2000 HP)
 - Added drop tables. last loot group was separated into a few different tables using inspiration from similar KSNM drop tables
 - Hell Flies can be charmed
 - Updated Spawn points
 - Reduced Hell Flies levels to be slightly lower than Beelzebub
 - Created skill list for bees to only use pollen and sharp sting
 - Increased size of Beezlebub to be bigger than the Hell Flies
 - Gave Beezlebub the ability to use benediction
 
 Evidence:
 https://youtu.be/Da25jMQSbZs
 https://ffxiclopedia.fandom.com/wiki/Infernal_Swarm?so=search

## Steps to test these changes

Enter KSNM Infernal Swarm in qu'bia arena to test

dallano committed:
KSNM Infernal Swarm

Basis and evidence received from the following:
https://ffxiclopedia.fandom.com/wiki/Infernal_Swarm
https://ffxi.allakhazam.com/db/quests.html?fquest=336

Changes:
Implemented KSNM Infernal Swarm
 - Gave Beelzebub a potent pollen ability (~2000 HP)
 - Added drop tables. last loot group was separated into a few different tables using inspiration from similar KSNM drop tables
 - Hell Flies can be charmed
 - Updated Spawn points
 - Reduced Hell Flies levels to be slightly lower than Beelzebub
 - Created skill list for bees to only use pollen and sharp sting
 - Increased size of Beezlebub to be bigger than the Hell Flies
 - Gave Beezlebub the ability to use benediction

zach2good commented on scripts/globals/mobskills/pollen.lua:
```diff
@@ -17,13 +17,24 @@ end
 
     potency = potency - math.random(0, potency/4)
 
+    if -- KSNM Infernal Swarm
```
You could just check against the name, instead of 3x hardcoded IDs?

zach2good commented on scripts/globals/mobskills/pollen.lua:
```diff
@@ -17,13 +17,24 @@ end
+        id == 17621258 or
+        id == 17621268
+    then
+        print("found ID")
```
Remove debug prints

zach2good commented on scripts/globals/mobskills/pollen.lua:
```diff
@@ -17,13 +17,24 @@ end
+        id == 17621268
+    then
+        print("found ID")
+        potency = 55
```
Got any sources for this?

dallano replied:
proof is linked in the commit. Should I also link in the PR comment?

dallano replied:
https://ffxi.allakhazam.com/db/quests.html?fquest=336#:~:text=He%20can%20also%20use%20Pollen%20as%20a%20TP%20move%20and%20recovers%20up%20to%202500%20HP.

zach2good replied:
A single forum post from 2005 doesn't do it for me I'm afraid. Hopefully, this is covered in a cap. I'm 70% sure that this KSNM is now uncapped on retail, so it wouldn't be hard to hop on and find out

dallano replied:
https://ffxiclopedia.fandom.com/wiki/Infernal_Swarm#:~:text=All%20bees%20have%20Pollen%20and%20Sharp%20Sting.%20Beelzebub%27s%20Pollen%20can%20heal%20for%20a%20couple%20thousand%20HP. is another source I originally linked in the commit

dallano replied:
After reviewing the live captures you linked me to, 2000 was definitely too powerful. Seemed to hover around ~700-~900. Fixed was pushed.

zach2good commented on scripts/zones/QuBia_Arena/mobs/Hell_Fly.lua:
```diff
@@ -0,0 +1,16 @@
+local entity = {}
+
+entity.onMobSpawn = function(mob)
+    mob:setMobMod(xi.mobMod.CHARMABLE, 1)
+    mob:setMod(xi.mod.SLEEPRES, 250)
```
Proof please

dallano replied:
proof is linked in the commit. Should I also link in the PR comment?

dallano replied:
It was beelzebub that has the sleep resistance. I pushed a commit correcting this.

zach2good replied:
No need to link proof in the code itself (unless its a very old and very well trusted theorycrafting forum post)

We just want to see where you got your information, because _SO MANY_ people have tried to pull a fast one on us with "what they remember", "what they think it is" etc. rather than collecting a bunch of different sources, using a cap, or making a cap from retail

dallano replied:
you got it. I'll be more rigorous in linking the basis on which I make the changes.

TeoTwawki replied:
FYI, sleepres doesn't go that high. stops at 100 last time I looked at how it was being used.

https://github.com/LandSandBoat/server/pull/2359/files/f6d4a98f0b4c2ea3ce92dbff43e965f664a2e6f0#diff-9c02459c581c8db660e64af625031e7f819e7cf6bf8dbdc08f3ff4527394b9e0

dallano replied:
After testing, 100 seems to be 10%. I can only receive immunity from the mob at 1000. Aerec and I spoke of this issue in voice earlier.

TeoTwawki replied:
@Xaver-DaRed did you rework stat res to 1000 scale in you magic rework and I missed it?

Xaver-DaRed replied:
I don't remember touching them. My rework hasnt reached status effects yet

dallano replied:
A quick test using 100 just now and I can still land everything

TeoTwawki replied:
I'm not seeing anything in our existing globals implying 1000 base, in fact I see code implying the 100 I expected.

> A quick test using 100 just now and I can still land everything

I'm not surprised, status res has never been retail correct and it may be that throwing massive numbers at it bumps up meva enough to be finally noticeable. It's not supposed to be just meva tho. 

https://www.bg-wiki.com/ffxi/Resist_Sleep
> - Causes the Sleep status effect to be fully resisted a certain percentage of the time.
> - At the moment, it is assumed that Resist Sleep +1 grants a 1% increase in this change to resist with an unknown cap. It is suspected the cap is lower than 100%.

https://ffxiclopedia.fandom.com/wiki/Resist_Sleep
> - Each upgrade of this trait further increases character's resistance.
> - This trait greatly shortens the duration of [sleep](https://ffxiclopedia.fandom.com/wiki/Sleep) type status effects by increasing the chance of half duration resists.
> - Gives a small chance to completely resist the effect of [sleep](https://ffxiclopedia.fandom.com/wiki/Sleep) status. When it procs the text will be preceeded by "Resist!" in a similar fashion to [Cover](https://ffxiclopedia.fandom.com/wiki/Cover).

TeoTwawki replied:
double checked how we are using them in items and existing mobs, it's all expecting 100 scaled in our trunk. if it somehow is not, and as I have seen no use case for fractions of a %, any place making it 1000 scale should probably be set to 100 rather than updating multiple dozens of mobs and items.

TeoTwawki replied:
@dallano for what it's worth I've complained the resist system was pretty much broken since about 2014 under old dsp. It's on our radar to redo in LSB but other things have taken priority. What I think is happening is your sleepres is just being added as flat meva.

dallano replied:
okay! I'll continue to give use the conventional scaling. You're right it would be ridiculous to code half the mobs as 1000 base just to fix the core issue and have to go back and change every mob lol

zach2good commented on scripts/zones/QuBia_Arena/npcs/Armoury_Crate.lua:
```diff
@@ -9,6 +9,116 @@ local entity = {}
+    [515] =
+    {
+        {
+            {itemid =   0, droprate = 350}, -- nothing
```
All these itemids should be tabled into items.lua

dallano replied:
You want me to add all the missing values into items.lua?

zach2good replied:
Yup

zach2good replied:
and then add those new values to this file: `itemid = xi.items. SCROLL_OF_SHELL_IV, ...`

dallano replied:
Should be good

zach2good commented on sql/bcnm_battlefield.sql:
```diff
@@ -801,6 +801,31 @@ INSERT INTO `bcnm_battlefield` VALUES (512,3,17621010,0); -- Ancient Warrior
 INSERT INTO `bcnm_battlefield` VALUES (512,3,17621011,0); -- Ancient Warrior
 INSERT INTO `bcnm_battlefield` VALUES (512,3,17621012,0); -- Ancient Warrior
 INSERT INTO `bcnm_battlefield` VALUES (512,3,17621013,0); -- Ancient Warrior
+INSERT INTO `bcnm_battlefield` VALUES (515,1,17621248,3); -- Infernal Swarm
```
All of these can be annotated with what they are, see the BCNM above

dallano replied:
Didn't annotate it like that as it was the only other BCNM I could find that was annotated that way. The rest declare the BCNM / KSNM once and list the values below. Is this the standard you want me to follow?

zach2good replied:
We want things to be as readable as possible, so if the name isn't built into the SQL entry itself, it should probably be added as a comment etc.

dallano replied:
alright you bet!

dallano commented:
@zach2good I'm not in the LSB. wasn't aware this existed. I'll check it out

dallano committed:
Styling fixes

Shortened check in pollen to check for name rather than ID
Fixed typo

zach2good commented:
It's a public discord: https://discord.gg/54r79GXc

Mortalelite commented:
>  I'm not in the LSB. wasn't aware this existed. I'll check it out

[Captures Discord](https://discord.gg/cJEpmGc) Incase didnt have the link

dallano committed:
Sleep res swap

Sleep res was incorrectly added to the hell fly instead of Beelzebub

dallano committed:
Styling fixes

Added more items from loot tables into items.lua
Removed unused variable from pollen.lua
Updated styling in bcnm_battlefield

TeoTwawki commented:
>  I'm not in the LSB discord. wasn't aware this existed. I'll check it out

well thats good coz we don't have one! we just make use of some community resources run by others that do :)

TeoTwawki commented on scripts/zones/QuBia_Arena/mobs/Beelzebub.lua:
```diff
@@ -0,0 +1,16 @@
+local entity = {}
+
+entity.onMobSpawn = function(mob)
+    mob:setMod(xi.mod.SLEEPRES, 250)
```
from your link:  
"Hell Fly can be charmed and slept, Beelzebub has a higher resistance to sleep.:"

That is not what I call proof that this should be 250. Even more so since status resists were 100 max last I checked and he's not supposed to be outright immune to sleep either.

dallano committed:
KSNM Infernal Swarm Fixes

Reduced pollen effect after reviewing live captures:
https://discord.com/channels/443544205206355968/443594542613004308/736051767057580042
https://www.youtube.com/watch?v=Da25jMQSbZs&ab_channel=BlargedyMcgee
Converted loot tables in armoury_crate to reference items.lua
Reduced the level of the hell flies to 70 after reviewing live captures:
https://discord.com/channels/443544205206355968/443594542613004308/736051767057580042
https://www.youtube.com/watch?v=Da25jMQSbZs&ab_channel=BlargedyMcgee
Gave Beelzebub sleep, silence, and paralyze immunity after reviewing live captures:
https://discord.com/channels/443544205206355968/443594542613004308/736051767057580042
https://www.youtube.com/watch?v=Da25jMQSbZs&ab_channel=BlargedyMcgee

dallano committed:
Update Beelzebub.lua

Fixing res values

TeoTwawki commented on scripts/globals/mobskills/pollen.lua:
```diff
@@ -24,6 +24,11 @@ mobskill_object.onMobWeaponSkill = function(target, mob, skill)
 
     potency = potency - math.random(0, potency/4)
 
+    if mob:getName() == "Beelzebub" then
```
A few things you should be aware of here:
- Pollen's whole methodology is fubar, we know it, and am not gonna make you try and sort that out! This is fine for now. [I have an open issue the subject of fubar mobskills that uses pollen as its example.](https://github.com/LandSandBoat/server/issues/2163) Someday we'll revisit.
- One thing I am asking to change is that we normally match on pool instead of on name. `if mob:getPool() == 385 then` in this case.
- there's [another PR making a similar change for Demonic Tiphia,](https://github.com/LandSandBoat/server/pull/2098/files#diff-8a71732b9e62411bd52ad3fc44663ed88abe4d68c907791c5bd61f70f033a1e7R29) whichever PR merges 1st will have no problem, the one that merges 2nd will have to update and handle the edit conflict.

dallano replied:
Okay, can do! not to throw zach under the bus, but I used getName() under his advisory
I'll definitely use this from now on.

TeoTwawki replied:
thats ok, I'm driving that bus and I know where he lives :) 



j/k zach!

dallano committed:
Update pollen.lua

Switch check from getName() to getPool()

dallano committed:
Update pollen.lua
